### PR TITLE
Add import status endpoint and progress options

### LINKED ISSUE #1850: Add historical data import endpoint
Screen Shot 2019-03-19 at 4 57 23 PM

The historical data import tool (above) requires endpoints for the following functionality:

_(It's probably best to split these items into separate PRs)_

- [ ] Updating the Customer and Order counts when a backfill date is chosen or "skip previously imported" is toggled
- [x] Initiating a historical data import #2034 
  * Params: # of days to backfill, whether to skip existing customers/orders
  * Persist the parameters for use in `WC_Admin_Reports_Sync` queries
- [ ] Querying the status of a historical data import
  * Status of current import if in progress - importing customers, importing orders
  * Date of oldest Order imported (used to show how much data has been imported)
- [x] Cancelling an in-process historical data import #2048
- [x] Deleting all previously imported data #2057 

See: p6riRB-3ZT-p2

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #1850

Adds an endpoint to query the current import progress and oldest date of import.  The date will be set to `-1` if `days` are not specific initially.

Also adds an endpoint to query potential totals prior to import.

### Thoughts/notes
* Opted to use options here instead of attempting to query pending/completed batches.  The latter seems like it would get messy quickly and is prone to inaccurate results with completed results.
* Orders total is not calculated until after the customers batch is finished since it's a dependent action.  If we want to show orders total while customers are being processed, we will probably need to create both customer and order batches initially so that we can calculate totals and set the orders batch to pending until the customers batch is complete.
* Imports should still be limited to one import at a time.  Will outline a separate issue to tackle this.

### Screenshots
Screen Shot 2019-05-08 at 3 56 39 PM

### Detailed test instructions:

1.  Start an import by posting to the `/wp-json/wc/v4/reports/import` endpoint.
2.  Send a get request to `/wp-json/wc/v4/reports/import/status`.
3.  Check that totals and progress count are reflected in the endpoint.
4.  Send a get request to `/wp-json/wc/v4/reports/import/totals` with `day` and `skip_existing` params.
5.  Make sure that total numbers returned are accurate.

### CODE REVIEW

File: includes/api/class-wc-admin-rest-reports-import-controller.php [hunk 1]
```diff
@@ -249,4 +261,27 @@ public function delete_imported_items( $request ) {
...
 	}
+
+	/**
+	 * Get the status of the current import.
+	 *
+	 * @param  WP_REST_Request $request Request data.
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function get_import_status( $request ) {
+		$delete = WC_Admin_Reports_Sync::delete_report_data();
```
Comment (Aljullu) [inline includes/api/class-wc-admin-rest-reports-import-controller.php]: I don't understand this line. Is it a leftover from another function?

**Review by Aljullu — COMMENTED** [review]
Thanks for working on this @joshuatf! Just a couple of comments/questions:

> Orders total is not calculated until after the customers batch is finished since it's a dependent action. If we want to show orders total while customers are being processed, we will probably need to create both customer and order batches initially so that we can calculate totals and set the orders batch to pending until the customers batch is complete.

According to the designs it looks like we should display the totals even before starting the import:

image

Not sure if that's possible or we should just hide the progress bars when the import hasn't started yet.

**Comment by joshuatf** [comment]
Made some changes to this PR:

* Added `is_importing` to the status route.
* Added a separate totals endpoint (testing instructions above) to query totals prior to importing.
* Set totals for current import prior to beginning import.
* Added tests for totals and status endpoints.

@Aljullu this is ready for another look if you have time.

File: includes/notes/class-wc-admin-notes.php [hunk 2]
```diff
@@ -131,7 +136,21 @@ public static function schedule_unsnooze_notes() {
...
+
+	/**
+	 * Clears all queued actions.
+	 */
+	public static function clear_queued_actions() {
+		$store = ActionScheduler::store();
+
+		if ( is_a( $store, 'WC_Admin_ActionScheduler_WPPostStore' ) ) {
+			// If we're using our data store, call our bespoke deletion method.
+			$store->clear_pending_wcadmin_actions();
```
Comment (Aljullu) [inline includes/notes/class-wc-admin-notes.php]: This method [seems to clear import actions](https://github.com/woocommerce/woocommerce-admin/blob/master/includes/class-wc-admin-actionscheduler-wppoststore.php#L42), right? Shouldn't we create a different method which only cleared note actions?

**Review by Aljullu — COMMENTED** [review]
Thanks for the last changes @joshuatf!

This is testing great and code looks good to me, I just added a small question in the code.

File: includes/notes/class-wc-admin-notes.php (hunk 2 above)
Comment (joshuatf) [reply includes/notes/class-wc-admin-notes.php]: Ya, I think you're right and we should actually extract the actions out of that method to make this reusable.  Updated in https://github.com/woocommerce/woocommerce-admin/pull/2189/commits/13753b7e47869c3b55ead67b34dad87757d99945

**Comment by joshuatf** [comment]
Thanks again for the review @Aljullu!  Added an update to separate out action clearing for notes and imports.

**Review by Aljullu — APPROVED** [review]
Thanks for the changes @joshuatf! LGTM

File: includes/class-wc-admin-actionscheduler-wppoststore.php [hunk 3]
```diff
@@ -36,23 +36,15 @@ protected function create_post_array( ActionScheduler_Action $action, DateTime $
...
-		$action_types = array(
-			WC_Admin_Reports_Sync::QUEUE_BATCH_ACTION,
-			WC_Admin_Reports_Sync::QUEUE_DEPEDENT_ACTION,
-			WC_Admin_Reports_Sync::CUSTOMERS_IMPORT_BATCH_ACTION,
-			WC_Admin_Reports_Sync::ORDERS_IMPORT_BATCH_ACTION,
-			WC_Admin_Reports_Sync::ORDERS_IMPORT_BATCH_INIT,
-			WC_Admin_Reports_Sync::SINGLE_ORDER_IMPORT_ACTION,
-		);
-
 		foreach ( $action_types as $action_type ) {
```
Comment (rrennick) [inline includes/class-wc-admin-actionscheduler-wppoststore.php]: Since this is a public function `$action_types` should probably be cast to an array.

**Pull request merged**